### ISSUE TITLE
The REST API stops working if `PSQLException` happens.

### ISSUE THREAD

**Issue by mir-am** [fasten-project/fasten#461]
On rare occasions, the Postgres DB connection may be disconnected and hence the REST API stops working due to `PSQLException`. For instance, this happens if the Postgres DB is restarted.
A fix is needed to kill the REST API's container in this case so that it'll be restarted with a new DB connection.

**Comment by MagielBruntink** [comment]
Instead of killing & restarting, could we figure out why this exception / the disconnect is happening? It doesn't seem limited to restarting the postgres DB. 

If you want only kill & restart, the REST API should crash on the PSQLException, not log and proceed. Kubernetes will then restart the process.

Here is an example log:
```
2022-05-30 08:36:27.393 DEBUG 1 --- [nio-8080-exec-9] o.s.web.servlet.DispatcherServlet        : GET "/packages/com.nryanov.genkai:genkai-cats_2.13/0.6.0-RC1/deps", parameters={}
2022-05-30 08:36:27.393 DEBUG 1 --- [nio-8080-exec-9] s.w.s.m.m.a.RequestMappingHandlerMapping : Mapped to eu.fasten.analyzer.restapiplugin.api.DependencyApi#getPackageDependencies(String, String, int, int, String, Long)
2022-05-30 08:36:27.394 DEBUG 1 --- [nio-8080-exec-9] o.s.web.servlet.DispatcherServlet        : Failed to complete request: org.jooq.exception.DataAccessException: SQL [select "public"."packages"."id", "public"."packages"."package_name", "public"."packages"."forge", "public"."packages"."project_name", "public"."packages"."repository", "public"."packages"."created_at" from "public"."packages" join "public"."package_versions" on "public"."packages"."id" = "public"."package_versions"."package_id" where (true and lower("public"."packages"."package_name") = lower(?) and lower("public"."package_versions"."version") = lower(?))]; This connection has been closed.
2022-05-30 08:36:27.396 ERROR 1 --- [nio-8080-exec-9] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is org.jooq.exception.DataAccessException: SQL [select "public"."packages"."id", "public"."packages"."package_name", "public"."packages"."forge", "public"."packages"."project_name", "public"."packages"."repository", "public"."packages"."created_at" from "public"."packages" join "public"."package_versions" on "public"."packages"."id" = "public"."package_versions"."package_id" where (true and lower("public"."packages"."package_name") = lower(?) and lower("public"."package_versions"."version") = lower(?))]; This connection has been closed.] with root cause

org.postgresql.util.PSQLException: This connection has been closed.
	at org.postgresql.jdbc.PgConnection.checkClosed(PgConnection.java:883) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.jdbc.PgConnection.prepareStatement(PgConnection.java:1744) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.postgresql.jdbc.PgConnection.prepareStatement(PgConnection.java:430) ~[postgresql-42.3.3.jar!/:42.3.3]
	at org.jooq.impl.ProviderEnabledConnection.prepareStatement(ProviderEnabledConnection.java:109) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.SettingsEnabledConnection.prepareStatement(SettingsEnabledConnection.java:82) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.AbstractResultQuery.prepare(AbstractResultQuery.java:210) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:307) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.AbstractResultQuery.fetchLazy(AbstractResultQuery.java:295) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.AbstractResultQuery.fetchLazyNonAutoClosing(AbstractResultQuery.java:316) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.SelectImpl.fetchLazyNonAutoClosing(SelectImpl.java:2867) ~[jooq-3.16.3.jar!/:na]
	at org.jooq.impl.ResultQueryTrait.fetchOne(ResultQueryTrait.java:506) ~[jooq-3.16.3.jar!/:na]
	at eu.fasten.core.data.metadatadb.MetadataDao.assertPackageExistence(Unknown Source) ~[core-0.0.8.jar!/:na]
	at eu.fasten.core.data.metadatadb.MetadataDao.getPackageDependencies(Unknown Source) ~[core-0.0.8.jar!/:na]
	at eu.fasten.analyzer.restapiplugin.api.DependencyApi.getPackageDependencies(Unknown Source) ~[classes!/:na]
	at jdk.internal.reflect.GeneratedMethodAccessor71.invoke(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Unknown Source) ~[na:na]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1067) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:963) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:655) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883) ~[spring-webmvc-5.3.18.jar!/:5.3.18]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:764) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53) ~[tomcat-embed-websocket-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117) ~[spring-web-5.3.18.jar!/:5.3.18]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:769) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:889) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1743) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) ~[tomcat-embed-core-9.0.60.jar!/:na]
	at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
```

**Comment by proksch** [comment]
What I have observed by working through SSH on several of the servers over the last weeks is that sometimes the connection is stuck for several seconds. This correlates with NFS errors in the syslog... I am not sure if we have a networking issue with the server, some kind of a hardware defect, or what else could go on there. In any way, right now I expect that the Postgres connection dies for the same reason.

**Comment by mir-am** [comment]
@MagielBruntink,
I also responded in the FASTEN mailing list.
As @proksch pointed out, this could be related to network issues. Also, our current Postgres configuration on `monster` should be investigated further.
For now, first, I'd propose to address this in the code and restart the container. Then, we should investigate our infrastructure like Postgres' configs, etc.

**Comment by proksch** [comment]
In my opinion, it does not make sense to randomly optimize configs and software if we are potentially facing hardware issues. It takes more than a misconfigured Postgres to make SSH unusable for multiple seconds, plus, such a temporary outage can also cause the Postgres exception. I personally expect that something more fundamental is going wrong.

I would propose that you try to investigate the system logs in a more structured way to get a better "overall picture" of the system in a moment when either such a "block", a bunch of NFS errors, or a `PSQLException` happens.... this will hopefully help us to spot correlations and pinpoint the actual problem (and I bet that all three happen at the same time).

**Comment by mir-am** [comment]
> In my opinion, it does not make sense to randomly optimize configs and software if we are potentially facing hardware issues. It takes more than a misconfigured Postgres to make SSH unusable for multiple seconds, plus, such a temporary outage can also cause the Postgres exception. I personally expect that something more fundamental is going wrong.
> 
> I would propose that you try to investigate the system logs in a more structured way to get a better "overall picture" of the system in a moment when either such a "block", a bunch of NFS errors, or a `PSQLException` happens.... this will hopefully help us to spot correlations and pinpoint the actual problem (and I bet that all three happen at the same time).

I totally agree that we/I need to look more closely at the sys logs. 
HOWEVER, for this particular issue, i.e., `PSQLException`, I am unsure if it's related to network issues. Both the Postgres server and the Java REST API are running on the same machine (`monster). As of now, the possible explanation is that the `monster` server was heavily loaded, which was the case when Paolo was running resource-intesntive experiments on `monster` last week. And this might have caused the Postgres server to close active connections.

**Comment by MagielBruntink** [comment]
The problem still happens now, however. I have restarted the restapi-java pod twice this morning and again it's down: https://api.fasten-project.eu/api/mvn/packages

I checked the configuration on monster in `/etc/postgres/13/main`: the Postgres server allows for up to 1000 connections. That is a lot, really. Did we allocate enough memory to support that?

It ma also be a matter of timeout happening; ie. queries taking longer (because big DB) than timeout allows, which then leads to a connection close.
